src/components/molecules/HistoricalChart/HistoricalChart.tsx

```diff
@@ -0,0 +1,353 @@
+    const legendArray = isMobile
+        ? ['H', 'L', 'Change']
+        : ['O', 'H', 'L', 'C', 'Change'];
+
+    const [legendData, setLegendData] = useState({
+        O: '',
+        H: '',
+        L: '',
+        C: '',
+        VOLUME_KEY_NAME: '',
+        Change: '',
+        usdVol: '',
+    });
+
+    const setupCharts = useCallback(
+        (candlestickSeries: TSeries, volumeSeries: TSeries) => {
+            volumeSeries.applyOptions({
+                baseLineVisible: false,
+            });
+
+            const candleData = data.map(item => ({
+                time: Number(item.time) as UTCTimestamp,
+                open: +item.open,
+                high: +item.high,
+                low: +item.low,
+                close: +item.close,
+            }));
+
+            const volumeData = data.map(item => ({
+                time: Number(item.time) as UTCTimestamp,
+                value: +item.vol,
+                color: item.open > item.close ? '#FF9FAE' : '#09A8B7',
+            }));
+
+            candlestickSeries.setData(candleData);
+            volumeSeries.setData(volumeData);
+        },
+        [data]
+    );
+
+    const handleDisplayLegend = (
+        candleData: WhitespaceData<Time> | CandlestickData<Time> | null,
+        volumeData: WhitespaceData<Time> | HistogramData<Time> | null
+    ) => {
+        const mergeData = {
+            ...candleData,
+            ...volumeData,
+        } as Omit<ITradingData, 'vol'> & { value: number };
+
+        const date = new Date(Number(candleData?.time) * 1000);
+        const formattedDate = `${date.getFullYear()}/${(
+            '0' +
+            (date.getMonth() + 1)
+        ).slice(-2)}/${('0' + date.getDate()).slice(-2)}`;
+
+        setHoverTime(formattedDate);
+
+        const isNegative = mergeData?.close < mergeData?.open;
+
+        setLegendData({
+            O: `${mergeData?.open.toFixed(2)}`,
+            H: `${mergeData?.high.toFixed(2)}`,
+            L: `${mergeData?.low.toFixed(2)}`,
+            C: `${mergeData?.close.toFixed(2)}`,
+            VOLUME_KEY_NAME: `${mergeData?.value}`,
+            Change: `${(
+                ((mergeData?.close - mergeData?.open) / mergeData?.open) *
+                100
+            ).toFixed(2)}%`,
+            usdVol: `${isNegative ? '-' : ''}${usdPrice * mergeData.value}`,
+        });
+    };
+
+    const subscribeToChartEvents = (
+        candleStickChart: IChartApi,
+        volumeChart: IChartApi,
+        candlestickSeries: TSeries,
+        volumeSeries: TSeries
+    ) => {
+        const updateLegendData = (param: MouseEventParams<Time>) => {
+            if (!param.seriesData.size) return;
+
+            let index = 0;
+            if (param.time) {
+                index = candlestickSeries
+                    .data()
+                    .findIndex(
+                        (item: Record<'time', Time>) => item.time === param.time
+                    );
+            }
+
```

Comment:
Please check if the index exists for valid data retrieval

Reply:
ok i added this